File: backend/server/services/chatbot.ts
import { storage } from '../storage.js';
import { azureAIService } from './azure-ai.js';
import { logger } from '../utils/logger.js';
import type { ChatRequest, Message } from '@shared/schema';

export interface ChatbotResponse {
  message: string;
  conversationId: number;
  messageId: number;
  timestamp: Date;
}

export class ChatbotService {
  async processMessage(request: ChatRequest, userId: number = 1): Promise<ChatbotResponse> {
    try {
      let conversationId = request.conversationId;
      
      // Create new conversation if none provided
      if (!conversationId) {
        const conversation = await storage.createConversation({
          title: this.generateConversationTitle(request.message),
          userId,
        });
        conversationId = conversation.id;
      }

      // Store user message
      const userMessage = await storage.createMessage({
        conversationId,
        content: request.message,
        isUser: true,
      });

      // Get conversation history for context
      const conversationHistory = await this.getConversationHistory(conversationId);
      
      // Search knowledge base for relevant context
      const contextResults = await azureAIService.searchKnowledgeBase(request.message);
      const context = contextResults.length > 0 ? contextResults.join('\n\n') : undefined;

      // Generate AI response
      const aiResponse = await azureAIService.chatCompletion({
        message: request.message,
        conversationHistory,
        context,
      });

      // Store AI response
      const botMessage = await storage.createMessage({
        conversationId,
        content: aiResponse.response,
        isUser: false,
      });

      // Update conversation timestamp
      await storage.updateConversation(conversationId, {
        updatedAt: new Date(),
      });

      logger.info(`Processed message for conversation ${conversationId}`, {
        userId,
        messageLength: request.message.length,
        responseLength: aiResponse.response.length,
      });

      return {
        message: aiResponse.response,
        conversationId,
        messageId: botMessage.id,
        timestamp: botMessage.timestamp,
      };
    } catch (error) {
      logger.error('Error processing chatbot message:', error);
      throw new Error('Failed to process message');
    }
  }

  private async getConversationHistory(conversationId: number): Promise<Array<{ role: 'user' | 'assistant'; content: string }>> {
    const messages = await storage.getConversationMessages(conversationId);
    
    // Get last 10 messages for context (excluding the current message)
    return messages
      .slice(-10)
      .map((msg) => ({
        role: msg.isUser ? 'user' as const : 'assistant' as const,
        content: msg.content,
      }));
  }

  private generateConversationTitle(message: string): string {
    // Generate a title from the first message
    const words = message.split(' ').slice(0, 6);
    let title = words.join(' ');
    
    if (message.length > title.length) {
      title += '...';
    }
    
    return title || 'New Conversation';
  }

  async getConversations(userId: number = 1): Promise<Array<{ id: number; title: string; updatedAt: Date }>> {
    const conversations = await storage.getUserConversations(userId);
    return conversations
      .sort((a, b) => b.updatedAt.getTime() - a.updatedAt.getTime())
      .map(conv => ({
        id: conv.id,
        title: conv.title,
        updatedAt: conv.updatedAt,
      }));
  }

  async getConversationMessages(conversationId: number): Promise<Message[]> {
    return await storage.getConversationMessages(conversationId);
  }
}

export const chatbotService = new ChatbotService();

File: backend/services/chatbot.py
"""
Chatbot service - Main business logic for chat operations
"""

from datetime import datetime
from typing import Dict, List, Any
from services.azure_ai import get_openai_service, get_search_service
from services.storage import get_storage_service
from schemas.chat import ChatMessageRequest, ChatMessageResponse
from utils.logger import get_logger

logger = get_logger(__name__)


class ChatbotService:
    """Main chatbot service handling message processing"""
    
    def __init__(self):
        self.openai_service = get_openai_service()
        self.search_service = get_search_service()
        self.storage = get_storage_service()
    
    async def process_message(self, request: ChatMessageRequest, user_id: int = 1) -> ChatMessageResponse:
        """Process a chat message and generate AI response"""
        
        try:
            conversation_id = request.conversation_id
            
            # Create new conversation if none provided
            if not conversation_id:
                conversation = await self.storage.create_conversation(
                    title=self._generate_conversation_title(request.message),
                    user_id=user_id
                )
                conversation_id = conversation["id"]
            
            # Store user message
            user_message = await self.storage.create_message(
                conversation_id=conversation_id,
                content=request.message,
                is_user=True
            )
            
            # Get conversation history for context
            conversation_history = await self._get_conversation_history(conversation_id)
            
            # Search knowledge base for relevant context
            context_results = await self.search_service.search_knowledge_base(request.message)
            context = "\n\n".join(context_results) if context_results else None
            
            # Generate AI response
            ai_response = await self.openai_service.chat_completion(
                message=request.message,
                conversation_history=conversation_history,
                context=context
            )
            
            # Store AI response
            bot_message = await self.storage.create_message(
                conversation_id=conversation_id,
                content=ai_response["response"],
                is_user=False
            )
            
            # Update conversation timestamp
            await self.storage.update_conversation(
                conversation_id=conversation_id,
                updated_at=datetime.utcnow()
            )
            
            logger.info(f"Processed message for conversation {conversation_id}")
            
            return ChatMessageResponse(
                message=ai_response["response"],
                conversation_id=conversation_id,
                message_id=bot_message["id"],
                timestamp=bot_message["timestamp"]
            )
            
        except Exception as e:
            logger.error(f"Error processing chatbot message: {e}")
            raise Exception("Failed to process message")
    
    async def get_conversations(self, user_id: int = 1) -> List[Dict[str, Any]]:
        """Get user conversations"""
        return await self.storage.get_user_conversations(user_id)
    
    async def get_conversation_messages(self, conversation_id: int) -> List[Dict[str, Any]]:
        """Get messages for a conversation"""
        return await self.storage.get_conversation_messages(conversation_id)
    
    async def _get_conversation_history(self, conversation_id: int) -> List[Dict[str, str]]:
        """Get conversation history for AI context"""
        messages = await self.storage.get_conversation_messages(conversation_id)
        
        # Get last 10 messages for context
        return [
            {
                "role": "user" if msg["is_user"] else "assistant",
                "content": msg["content"]
            }
            for msg in messages[-10:]
        ]
    
    def _generate_conversation_title(self, message: str) -> str:
        """Generate a title from the first message"""
        words = message.split()[:6]
        title = " ".join(words)
        
        if len(message) > len(title):
            title += "..."
        
        return title or "New Conversation"


# Global service instance
_chatbot_service = None


def get_chatbot_service() -> ChatbotService:
    """Get chatbot service singleton"""
    global _chatbot_service
    if _chatbot_service is None:
        _chatbot_service = ChatbotService()
    return _chatbot_service